src/parquet/file_reader.cc

```diff
@@ -0,0 +1,764 @@
+    }
+    const format::ColumnChunk& column_chunk = metadata().row_groups[row_group].columns[column];
+    const schema::primitive_node& leaf = *schema().leaves[column];
+    return [this, &column_chunk] {
+        if (!column_chunk.__isset.file_path) {
+            return seastar::make_ready_future<seastar::file>(file());
+        } else {
+            return seastar::open_file_dma(path() + column_chunk.file_path, seastar::open_flags::ro);
+        }
+    }().then([&column_chunk, &leaf] (seastar::file f) {
+        return [&column_chunk, f] {
+            if (column_chunk.__isset.meta_data) {
+                return seastar::make_ready_future<std::unique_ptr<format::ColumnMetaData>>(
+                        std::make_unique<format::ColumnMetaData>(column_chunk.meta_data));
+            } else {
+                return read_chunk_metadata(seastar::make_file_input_stream(f, column_chunk.file_offset));
+            }
+        }().then([f, &leaf] (std::unique_ptr<format::ColumnMetaData> column_metadata) {
+            size_t file_offset = column_metadata->__isset.dictionary_page_offset
+                                 ? column_metadata->dictionary_page_offset
+                                 : column_metadata->data_page_offset;
+            page_reader page_reader{
+                    seastar::make_file_input_stream(f, file_offset, column_metadata->total_compressed_size)};
+            return column_chunk_reader<T>(leaf, std::move(page_reader), column_metadata->codec);
+        });
+    });
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+template class column_chunk_reader<format::Type::INT32>;
+template class column_chunk_reader<format::Type::INT64>;
+template class column_chunk_reader<format::Type::INT96>;
+template class column_chunk_reader<format::Type::FLOAT>;
+template class column_chunk_reader<format::Type::DOUBLE>;
+template class column_chunk_reader<format::Type::BOOLEAN>;
+template class column_chunk_reader<format::Type::BYTE_ARRAY>;
+template class column_chunk_reader<format::Type::FIXED_LEN_BYTE_ARRAY>;
+/*
+ * Explicit instantiation of value_decoder shouldn't be needed,
+ * because column_chunk_reader<T> has a value_decoder<T> member.
+ * Yet, without explicit instantiation of value_decoder<T>,
+ * value_decoder<T>::read_batch is not generated. Why?
+ *
+ * Quote: When an explicit instantiation names a class template specialization,
+ * it serves as an explicit instantiation of the same kind (declaration or definition)
+ * of each of its non-inherited non-template members that has not been previously
+ * explicitly specialized in the translation unit. If this explicit instantiation
+ * is a definition, it is also an explicit instantiation definition only for
+ * the members that have been defined at this point.
+ * https://en.cppreference.com/w/cpp/language/class_template
+ *
+ * Has value_decoder<T> not been "defined at this point" or what?
```

Comment:
Can't help here, but it's quite possible you'll get your answers once these patches are sent to seastar-dev.